### ISSUE TITLE
OOF hill climbing blend weight optimizer

### ISSUE THREAD

**Issue by haukesteffen** [haukesteffen/TabularShenanigans#222]
## Goal

Add `scripts/optimize_blend_weights.py` — a post-sweep analysis script that finds optimal ensemble weights via hill climbing on OOF predictions, then emits a ready-to-paste `config.yaml` blend candidate snippet.

## Motivation

After the s6e3 sweep produces 192 trained candidates, the next leaderboard gain comes from ensembling. A naive top-N-by-CV-AUC blend ignores correlation between candidates — two nearly identical XGBoost models contribute less than a diverse LightGBM+XGBoost pair even if the XGBoost pair scores higher individually. Hill climbing on OOF predictions naturally handles this: a redundant candidate simply fails to clear the improvement threshold and gets weight 0.

## Algorithm

1. Start: best single candidate (by `cv_score_mean`) gets weight 1.0, all others 0.0
2. Each round: for every candidate `i`, try `weight[i] += step_size` and `weight[i] -= step_size`
   - Clip weights to `[0, ∞)`, re-normalize to sum=1
   - Evaluate blended OOF AUC (`roc_auc_score(y_true, weighted_avg_oof)`)
   - Accept if improvement exceeds `threshold`
3. After a full round with no accepted perturbation, halve `step_size`
4. Stop when `step_size < 1e-6`

Candidates with weight 0 throughout are naturally excluded from the output snippet.

### 1. Config and MLflow setup
- Follow the pattern in `verify_sweep_execution.py`: load `config.yaml` with `yaml.safe_load`, extract `tracking_uri`, `competition_slug`, and CV config (`n_splits`, `shuffle`, `random_state`) and `primary_metric`
- Set MLflow tracking URI, get experiment by competition slug

### 2. Candidate loading
- Query MLflow for all `FINISHED` runs with `tags.run_kind = 'candidate'` and `tags.candidate_type = 'model'`
- Filter in Python to runs where `cv__n_splits`, `cv__shuffle`, `cv__random_state` match config — mismatched fold structures produce incompatible OOF rows
- Read `cv_score_mean` from each run's metrics for the starting sort
- Report how many candidates were found and how many passed the CV filter

### 3. OOF download
- For each qualifying run, use `client.download_artifacts(run_id, "candidate/oof_predictions.csv", dst_path)` into a temp dir — download only this file, not the full bundle
- Load each CSV, sort by `row_idx`, extract `y_pred` (float probabilities) and `y_true` (from first candidate, validated to match across all others)
- Stack into an `(N_candidates, N_rows)` OOF matrix

### 4. Hill climbing
- Initialize weights: `[1.0, 0.0, ..., 0.0]` (best candidate leads)
- Each perturbation: copy weights, add/subtract `step_size` at index `i`, clip to `[0, ∞)`, normalize, score
- Accept if `new_score - best_score > threshold`
- Halve `step_size` after a round with no improvement; stop at `step_size < 1e-6`

### 5. Output
```
Found 192 FINISHED model candidates. 189 match current CV config (7-fold, shuffle=True, seed=42).

Optimization trace:
  start (equal weights, 189 components): OOF AUC = 0.8461
  round 1: accepted 4 perturbations, AUC = 0.8468
  round 2: accepted 1 perturbation,  AUC = 0.8471
  ...
  converged after 14 rounds. Final OOF AUC = 0.8473

Components (12 of 189 have non-zero weight):
  rank  candidate_id                               model        recipe  cv_auc   weight
     1  fr2_lightgbm_median_onehot                 lightgbm     fr2     0.8452   0.312
     2  fr3_xgboost_standardize_ordinal            xgboost      fr3     0.8449   0.188
     ...

--- paste into config.yaml candidates: ---
- candidate_type: blend
  base_candidate_ids:
    - fr2_lightgbm_median_onehot
    - fr3_xgboost_standardize_ordinal
    ...
  weights:
    - 0.312
    - 0.188
    ...
```

### 6. CLI
```
uv run python scripts/optimize_blend_weights.py [--step-size 0.05] [--threshold 1e-5]
```

## User workflow after running the script
1. Paste the config snippet into `config.yaml`
2. `ts train` → creates the blend MLflow run
3. `ts submit`

## Acceptance criteria
- [ ] Script runs to completion after the s6e3 sweep finishes
- [ ] Candidate pool filtered to FINISHED model runs with matching CV config
- [ ] Only `oof_predictions.csv` downloaded per candidate (not full bundle)
- [ ] Hill climbing terminates cleanly; output shows optimization trace
- [ ] Config snippet is valid YAML and accepted by `ts train` without modification
- [ ] Manually verified: blended OOF AUC exceeds best individual candidate AUC
